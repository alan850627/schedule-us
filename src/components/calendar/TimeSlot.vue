<template>
  <div class="time-slot noselect" :class="status" v-on:mouseover="updateStatusDrag" v-on:mousedown="updateStatusClick">
    {{ displayTime }}
  </div>
</template>

<script>
import moment from 'moment'

export default {
  name: 'TimeSlot',

  components: {
  },

  props: {
    mouseDown: {
      type: Boolean,
      default: false
    },
    username: {
      type: String,
      default: ''
    },
    viewSingle: {
      type: Boolean,
      default: false
    },
    editable: {
      type: Boolean,
      default: false
    },
    startTime: {
      required: true,
      type: Number,
      default: 0
    },
    length: {
      type: Number,
      default: 3600000
    },
    response: {
      type: Object,
      default: () => { return {} }
    }
  },

  computed: {
    displayTime: function () {
      return moment(this.startTime).format('LT')
    },

    status: function () {
      let usernames = Object.keys(this.response)
      if (usernames.length === 0) {
        return 'no'
      }

      if (this.viewSingle) {
        return this.response[this.username]
      }

      let status = 'yes'
      usernames.forEach(username => {
        if (this.response[username] === 'maybe') {
          status = 'maybe'
        } else if (this.response[username] === 'no') {
          return 'bad'
        }
      })
      return status
    }
  },

  data () {
    return {
      changeFlag: false
    }
  },

  methods: {
    updateStatusClick: function () {
      if (this.editable) {
        this.cycleStatus()
      }
    },
    updateStatusDrag: function () {
      if (!this.changeFlag && this.editable && this.mouseDown) {
        this.cycleStatus()
      }
    },
    cycleStatus: function () {
      if (!this.response[this.username] || this.response[this.username] === 'no') {
        this.$set(this.response, this.username, 'yes')
      } else if (this.response[this.username] === 'yes') {
        this.$set(this.response, this.username, 'maybe')
      } else if (this.response[this.username] === 'maybe') {
        this.$set(this.response, this.username, 'no')
      }
      this.changeFlag = true
    }
  },

  watch: {
    mouseDown: function (newMouseDown, oldMouseDown) {
      if (!newMouseDown) {
        this.changeFlag = false
      }
    }
  },

  mounted () {
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.yes {
  background:#D4EDDA
}

.no {
  background:white
}

.bad {
  background:#F7D7DA
}

.maybe {
  background:#FFF3CD
}

.noselect {
  -webkit-touch-callout: none; /* iOS Safari */
    -webkit-user-select: none; /* Safari */
     -khtml-user-select: none; /* Konqueror HTML */
       -moz-user-select: none; /* Firefox */
        -ms-user-select: none; /* Internet Explorer/Edge */
            user-select: none; /* Non-prefixed version, currently
                                  supported by Chrome and Opera */
}
</style>
