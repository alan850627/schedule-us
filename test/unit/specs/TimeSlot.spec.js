import TimeSlot from '@/components/calendar/TimeSlot'
import { shallow } from 'vue-test-utils'

describe('TimeSlot.vue', () => {
  it('has the expected html structure', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'test-user',
        startTime: 1524718800000
      }
    })
    expect(cmp.element).toMatchSnapshot()
  })
  it('comes with default background', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'test-user',
        startTime: 1524718800000
      }
    })
    expect(cmp.find('.time-slot').hasStyle('background', 'white')).toBe(true)
  })
  it('display correct time', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'test-user',
        startTime: 1524718800000
      }
    })
    expect(cmp.vm.displayTime).toBe('12:00 AM')
  })
  it('display correct hover text', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'test-user',
        startTime: 1524718800000,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user3': 'maybe',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    expect(cmp.vm.hoverText).toBe('[yes]- user1; user4; \n[maybe]- user3; user6; \n[no]- user2; user5; ')
  })
  it('display correct color', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'user1',
        startTime: 1524718800000,
        viewSingle: true,
      }
    })
    expect(cmp.vm.status).toBe('white')
  })
  it('display correct color', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'user1',
        startTime: 1524718800000,
        viewSingle: true,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user3': 'maybe',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    expect(cmp.vm.status).toBe('rgb(0, 230, 0)')
  })
  it('display correct color', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'user2',
        startTime: 1524718800000,
        viewSingle: true,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user3': 'maybe',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    expect(cmp.vm.status).toBe('white')
  })
  it('display correct color', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'user3',
        startTime: 1524718800000,
        viewSingle: true,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user3': 'maybe',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    expect(cmp.vm.status).toBe('rgb(230, 230, 0)')
  })
  it('display correct color', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'user3',
        startTime: 1524718800000,
        viewSingle: false,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user3': 'maybe',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    expect(cmp.vm.status).toBe('rgb(153.33333333333334, 153.33333333333334, 0)')
  })
  it('user clicks cycle status', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'user3',
        startTime: 1524718800000,
        viewSingle: true,
        editable: true,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    cmp.vm.updateStatusClick()
    expect(cmp.vm.response['user3']).toBe('yes')
    cmp.vm.changeFlag = false
    cmp.vm.updateStatusClick()
    expect(cmp.vm.response['user3']).toBe('maybe')
    cmp.vm.changeFlag = false
    cmp.vm.updateStatusClick()
    expect(cmp.vm.response['user3']).toBe('no')
    expect(cmp.vm.changeFlag).toBe(true)
  })
  it('user drags cycle status', () => {
    let cmp = shallow(TimeSlot, {
      propsData: {
        username: 'user3',
        startTime: 1524718800000,
        viewSingle: true,
        editable: true,
        mouseDown: true,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    cmp.vm.updateStatusDrag()
    expect(cmp.vm.response['user3']).toBe('yes')
  })
  it('mouse down watcher', next => {
    let cmp = shallow(TimeSlot, {
      data: {
        changeFlag: true
      },
      propsData: {
        username: 'user3',
        startTime: 1524718800000,
        viewSingle: true,
        editable: true,
        mouseDown: true,
        response: {
          'user1': 'yes',
          'user2': 'no',
          'user4': 'yes',
          'user5': 'no',
          'user6': 'maybe'
        }
      }
    })
    cmp.vm.mouseDown = false
    cmp.vm.$nextTick(() => {
      expect(cmp.vm.changeFlag).toBe(false)
      next()
    })
  })
})
