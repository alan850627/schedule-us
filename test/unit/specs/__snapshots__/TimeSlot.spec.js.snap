// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeSlot.vue has the expected html structure 1`] = `
<div
  class="time-slot noselect"
  id="timeslot"
  placement="rightbottom"
  style="background: white;"
  title="[yes]- 
[maybe]- 
[no]- "
>
  
  12:00 AM

</div>
`;
